From 2e26428dc4d450e974ceb9bc737f691057cd80b6 Mon Sep 17 00:00:00 2001
From: ywc689 <ywc689@163.com>
Date: Fri, 28 Jun 2019 17:02:40 +0800
Subject: [PATCH 3/3] driver:kni: enable flow_item type comparsion in
 flow_fdir_cmp

the existence is checked before adding/deleting a fdir flow, but
the flow type is not compared in 'flow_fdir_cmp', which resulting
in the failure or unwanted behavior in adding/deleting two same
fdir flows with flow type(such as ipv4 tcp/udp) different only.
---
 drivers/net/mlx5/mlx5_flow.c | 5 +++++
 1 file changed, 5 insertions(+)

diff --git a/drivers/net/mlx5/mlx5_flow.c b/drivers/net/mlx5/mlx5_flow.c
index 222cd81..d99edce 100644
--- a/drivers/net/mlx5/mlx5_flow.c
+++ b/drivers/net/mlx5/mlx5_flow.c
@@ -2668,6 +2668,11 @@ struct rte_flow *
 static int
 flow_fdir_cmp(const struct mlx5_fdir *f1, const struct mlx5_fdir *f2)
 {
+	unsigned i;
+	for (i = 0; i < sizeof(f1->items)/sizeof(f1->items[0]); i++) {
+		if (f1->items[i].type != f2->items[i].type)
+			return 1;
+	}
 	if (FLOW_FDIR_CMP(f1, f2, attr) ||
 	    FLOW_FDIR_CMP(f1, f2, l2) ||
 	    FLOW_FDIR_CMP(f1, f2, l2_mask) ||
-- 
1.8.3.1

